### PR TITLE
feat(appset): Implement additional values and their interpolation for Git and SCM generators (#10751)

### LINKED ISSUE #10751: Allow values to be added & interpolated in Git and SCM generators
# Summary

After reading #9978 , I figured it would be nice to extend the functionality to set and interpolate `values` from the cluster ApplicationSet generator to other generators as well.

# Motivation

I have multiple clusters and multiple applications configured in their own repositories. I want these applications to be able to deploy themselves to clusters by having a dedicated branch per cluster.

One option is to use the cluster generator to discover configured clusters. Assuming those have a `metadata.labels.location` set, the SCM generator could discover those branches.

To combine those two, I could then have a `values.selector` label. The cluster generator sets `values.selector = metadata.labels.location` while the SCM generator does `values.selector = branch`. This then can be used by the merge generator to combine settings for each.

# Proposal

Re-use the already existing interpolation from the cluster generator and do the same for other interesting generators.

### PR DESCRIPTION
**Pull request by girstenbrei**
Closes #10751

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: applicationset/generators/value_interpolation.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package generators
+
+import (
+	"fmt"
+)
+
+func appendTemplatedValues(values map[string]string, params map[string]interface{}, useGoTemplate bool) error {
+	// We create a local map to ensure that we do not fall victim to a billion-laughs attack. We iterate through the
+	// cluster values map and only replace values in said map if it has already been whitelisted in the params map.
```
Comment (crenshaw-dev) [inline applicationset/generators/value_interpolation.go:9]: Could we go ahead and change "whitelisted" to "allowlisted"?

File: docs/operator-manual/applicationset/Generators-Merge.md [hunk 2]
```diff
@@ -111,6 +111,50 @@ When merged with the updated base parameters, the `values.redis` value for the p
...
+                selector: '{{ metadata.labels.location }}'
+          # The git repo may have different branches which correspond to the
+          # cluster locations, using these as a selector.
+          - git:
+              repoURL: https://github.com/argoproj/argocd-example-apps/
+              revision: HEAD
+              directories:
+              - path: '*'
+              values:
+                cluster: '{{ branch }}'
```
Comment (crenshaw-dev) [inline docs/operator-manual/applicationset/Generators-Merge.md:142]: Should `cluster` here be `selector` instead?

File: docs/operator-manual/applicationset/Generators-SCM-Provider.md [hunk 3]
```diff
@@ -154,21 +154,7 @@ metadata:
...
-        # Credentials for Basic authentication. Required for private repositories.
-        basicAuth:
-          # The username to authenticate with
-          username: myuser
-          # Reference to a Secret containing the password or personal access token.
-          passwordRef:
-            secretName: mypassword
-            key: password
-        # Support for filtering by labels is TODO. Bitbucket server labels are not supported for PRs, but they are for repos
+      cccccccccccccccccccccc
```
Comment (crenshaw-dev) [inline docs/operator-manual/applicationset/Generators-SCM-Provider.md:157]: Accidental delete?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
I'm very excited about this, thanks for the PR! Just had a few comments after a first pass review.

File: applicationset/generators/scm_provider.go [hunk 4]
```diff
@@ -148,9 +148,16 @@ func (g *SCMProviderGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha
 			"short_sha":        repo.SHA[:shortSHALength],
 			"labels":           strings.Join(repo.Labels, ","),
 			"branchNormalized": utils.SanitizeName(repo.Branch),
-		})
+		}
+
+		err := appendTemplatedValues(appSetGenerator.SCMProvider.Values, params, applicationSetInfo.Spec.GoTemplate)
+		if err != nil {
+			return nil, err
```
Comment (crenshaw-dev) [inline applicationset/generators/scm_provider.go:155]: Wrap error?

File: applicationset/generators/value_interpolation.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	// Once we iterate through all the cluster values we can then safely merge the `tmp` map into the main params map.
+	tmp := map[string]interface{}{}
+
+	for key, value := range values {
+		result, err := replaceTemplatedString(value, params, useGoTemplate)
+
+		if err != nil {
+			return err
```
Comment (crenshaw-dev) [inline applicationset/generators/value_interpolation.go:17]: Wrap error?

File: applicationset/generators/value_interpolation.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+		params[key] = value
+	}
+
+	return nil
+}
+
+func replaceTemplatedString(value string, params map[string]interface{}, useGoTemplate bool) (string, error) {
+	replacedTmplStr, err := render.Replace(value, params, useGoTemplate)
+	if err != nil {
+		return "", err
```
Comment (crenshaw-dev) [inline applicationset/generators/value_interpolation.go:40]: Wrap error?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Mostly repetitive, nit-picky stuff.  :-)

**Comment by crenshaw-dev** [comment]
@girstenbrei separate is fine! I'll squash on merge.  :-)

**Comment by crenshaw-dev** [comment]
@girstenbrei can you merge master and run codegen again? There was a bug preventing CRD updates for ApplicationSet.

**Comment by girstenbrei** [comment]
@crenshaw-dev Done, saw changes in code generated for ApplicationSet, but I'm unable to judge if these are now 'correct' :grin: 
Let me know if you need anything else to merge!

**Comment by girstenbrei** [comment]
Resolved, @crenshaw-dev , apparently someone else wanted to wrap those errors :grin:

**Comment by girstenbrei** [comment]
Merged master, resolving conflict in auto-generated code

**Comment by girstenbrei** [comment]
Hey there, I did merge from master without additional changes to be up to date, now a few of those integration tests fail. The don't seem to be related to the changes and I don't know how to restart those tests (maybe a one-time-failure?).
Could you give me some advice how to proceed?

**Comment by crenshaw-dev** [comment]
@girstenbrei it's not you! A Recent git update broke some of our tests. I'll merge master into this branch after fixing CI.

**Comment by girstenbrei** [comment]
Greets, I resolved a MC in the auto-generated files and updated the PR. Anything I can do to support merging?

**Comment by girstenbrei** [comment]
@crenshaw-dev Greets! I don't want to nag, so sorry for asking again: Anything I can do to support merging? Is there any timeline on how long it might take?

**Comment by crenshaw-dev** [comment]
@girstenbrei apologies for the silence, I had to go all-in on some high priority 2.6 features.

I really want this feature and am assigning it to myself for the 2.7 release. Will return with feedback ASAP.

**Comment by girstenbrei** [comment]
@crenshaw-dev  Thank you for making 2.6 happen and thank you for your community engagement!

I tried to merge in upstream, especially after #10523 . Unfortunately I now have a rather cryptic (at least to me) message regarding codegen at the `go run ./hack/gen-catalog docs` step (loads of `pkg/apiclient/applicationset/applicationset.pb.go:645:34: ambiguous selector m.Applicationset.MarshalToSizedBuffer`). I'm not sure what that means, will try to figure it out within the next days (not sure, It might relate to the change in `applicationset_types`), even the Lint CI step shows it. 

Greets,
Chris

File: docs/operator-manual/applicationset/Generators-Merge.md [hunk 7]
```diff
@@ -111,6 +111,68 @@ When merged with the updated base parameters, the `values.redis` value for the p
...
+kind: ApplicationSet
+metadata:
+  name: cluster-git
+spec:
+  generators:
+    # merge 'parent' generator:
+    # Use the selector set by both child generators to combine them.
+    - merge:
+        mergeKeys:
+          - values.selector
```
Comment (crenshaw-dev) [inline docs/operator-manual/applicationset/Generators-Merge.md:130]: @girstenbrei does this work for go-templated appsets? I think in that case `values` is a proper object. I'm not sure whether this key will behave.

Comment (girstenbrei) [reply docs/operator-manual/applicationset/Generators-Merge.md:130]: Hi @crenshaw-dev , good catch: Just tested this, indeed, with go-template enabled this does not work (the merge key resolves to "values.selector" and not ~"values"."selector". This seems like a current limitation of the merge generator; using go-templating the generator is not able to merge on a nested key.

If that's ok with you, I'd adapt the documentation to reflect this and open another issue regarding the merge operator supporting nested keys, keeping the rest of this PR as is?

Comment (crenshaw-dev) [reply docs/operator-manual/applicationset/Generators-Merge.md:130]: Makes sense, thanks!

**Comment by wanddynosios** [comment]
@girstenbrei Thanks for the feature! It would be very useful for us, too.

Can we help in any way? / What is missing to proceed here?

**Comment by crenshaw-dev** [comment]
Finally got a chance to come back to this! I'm rebasing now.

**Review by crenshaw-dev — APPROVED** [review]

**Pull request merged**